### PR TITLE
Add support for configVersion 2.

### PR DESCRIPTION
**Pull request by andrew-vant**
Version 2 is somewhat flatter, noticeably cleaner, and allows for both
yaml and json input files.

I've been talking with Bryan about this over the last few days. For an idea of why it's useful, compare the readability of the [old sample config](https://github.com/rackerlabs/curryproxy/blob/master/etc/routes.sample.json) to the [new one](https://github.com/andrew-vant/curryproxy/blob/config_v2/etc/routes.sample.yaml). Probably shouldn't merge this as-is; I haven't updated tests, etc.

This replaces basically everything in route_factory.py; I didn't remove the parse functions (yet) because it made the diff difficult to read, but nothing references them anymore.

### CODE REVIEW

**Comment by andrew-vant** [comment]
Just for the record, I know the tests are going to fail; I put in the PR mostly because it's a convenient way to demonstrate what I'm trying to do and get some feedback.

File: curryproxy/routes/route_factory.py [hunk 1]
```diff
@@ -56,6 +60,126 @@ def parse_ignore_rules(route_config):
...
+
+        yield EndpointsRoute(**route)
+
+
+def _convert(route_configs):
+    """Update old configuration data structure versions.
+
+    Returns a new data structure of the current version.
+
+    Should this be public?"""
```
Comment (BryanSD) [inline curryproxy/routes/route_factory.py]: Nah, and we can remove the comment now.

File: curryproxy/routes/route_factory.py [hunk 2]
```diff
@@ -56,6 +60,126 @@ def parse_ignore_rules(route_config):
...
+
+    # Endpoint routes had one key renamed to match the EPR constructor, and the
+    # endpoints subdict was changed from a list of key-value dicts to a flat
+    # dict (which also matches the constructor). It's cleaner here to build the
+    # list first and modify it afterward.
+    newconf['routes'] = [d for d in route_configs
+                         if 'endpoints' in d]
+
+    for route in newconf['routes']:
+        route['url_patterns'] = [route.pop('route')]
```
Comment (BryanSD) [inline curryproxy/routes/route_factory.py]: In the v1 and v2 samples, this is only a string, not an array. Maybe I'm missing something?

File: curryproxy/routes/route_factory.py [hunk 3]
```diff
@@ -56,6 +60,126 @@ def parse_ignore_rules(route_config):
...
+        # Convert error ranges into discrete lists.
+        ignore = []
+        for err in route['ignore_errors']:
+            err = str(err)
+            bounds = err.split("-")
+            start, end = int(bounds[0]), int(bounds[-1])
+            ignore.extend(range(start, end+1))
+        route['ignore_errors'] = ignore
+
+        # Accept singular url_pattern or plural url_patterns.
```
Comment (BryanSD) [inline curryproxy/routes/route_factory.py]: Why do we need this bit?

File: curryproxy/routes/route_factory.py [hunk 4]
```diff
@@ -56,6 +60,126 @@ def parse_ignore_rules(route_config):
...
+
+    for route in newconf['routes']:
+        route['url_patterns'] = [route.pop('route')]
+        route['endpoints'] = {ep['id']: ep['url']
+                              for ep in route['endpoints']}
+
+    return newconf
+
+
 def parse_dict(route_config):
```
Comment (BryanSD) [inline curryproxy/routes/route_factory.py]: Do we use this function anymore after this change?

Comment (andrew-vant) [inline curryproxy/routes/route_factory.py]: No. As mentioned in the initial comment, I only left it there because removing it made the diff unreadable. It can be yanked in a separate PR.

File: curryproxy/routes/route_factory.py (hunk 3 above)
Comment (andrew-vant) [inline curryproxy/routes/route_factory.py]: We don't need it, it's just convenient.

File: curryproxy/routes/route_factory.py (hunk 1 above)
Comment (andrew-vant) [inline curryproxy/routes/route_factory.py]: Fair enough.

File: curryproxy/routes/route_factory.py (hunk 2 above)
Comment (andrew-vant) [inline curryproxy/routes/route_factory.py]: If I remember right, the EndpointsRoute constructor expects url_patterns to be an array. The dict this builds eventually gets splatted into the constructor args at the bottom of load().

File: curryproxy/routes/route_factory.py (hunk 4 above)
Comment (BryanSD) [inline curryproxy/routes/route_factory.py]: Ah, yes, I forgot it was mentioned in the PR description. I'd still like to remove it in this PR before merging so we have one unit of work taking care of everything.

File: curryproxy/routes/route_factory.py (hunk 2 above)
Comment (BryanSD) [inline curryproxy/routes/route_factory.py]: Ah, yes I see. This makes sense, and I think nullifies all my other comments about `url_patterns`. Just looks like a couple of easy changes then and this PR is good-to-go after getting a few more eyes on it.

**Comment by andrew-vant** [comment]
Superceded by #55
